src/zarr/store/memory.py

```diff
@@ -101,3 +101,63 @@ async def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
             for key in self._store_dict:
                 if key.startswith(prefix + "/") and key != prefix:
                     yield key.removeprefix(prefix + "/").split("/")[0]
+
+
+class GpuMemoryStore(MemoryStore):
+    _store_dict: MutableMapping[str, Buffer]
+
+    def __init__(
+        self, store_dict: MutableMapping[str, Buffer] | None = None, *, mode: OpenMode = "r"
+    ):
+        super().__init__(mode=mode)
+        self._store_dict = {}
+        if store_dict:
+            self._store_dict = {k: GpuBuffer.from_buffer(store_dict[k]) for k in iter(store_dict)}
+
+    def __str__(self) -> str:
+        return f"gpumemory://{id(self._store_dict)}"
+
+    def __repr__(self) -> str:
+        return f"GpuMemoryStore({str(self)!r})"
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: tuple[int | None, int | None] | None = None,
+    ) -> Buffer | None:
+        assert isinstance(key, str)
+        try:
+            value = self._store_dict[key]
+            start, length = _normalize_interval_index(value, byte_range)
+            return value[start : start + length]
+        except KeyError:
+            return None
+
+    async def get_partial_values(
+        self,
+        prototype: BufferPrototype,
+        key_ranges: list[tuple[str, tuple[int | None, int | None]]],
+    ) -> list[Buffer | None]:
+        # All the key-ranges arguments goes with the same prototype
+        async def _get(key: str, byte_range: tuple[int, int | None]) -> Buffer | None:
+            # Q: use prototype here to convert to bespoke buffer class? If so, how?
+            return await self.get(key, prototype=prototype, byte_range=byte_range)
+
+        vals = await concurrent_map(key_ranges, _get, limit=None)
+        return vals
+
```

Comment:
With #1974, `get` and `get_partial_values` from `MemoryStore` should automatically return the buffer type specified in the `prototype` argument.

```suggestion
```

Reply:
This is a good point. Maybe worth adding some tests against a non-default prototype to ensure this is properly happening for all stores?

Btw, in #1974, how would the `from_buffer` work if you pass in a `GpuBuffer`? Won't that just set `_data` to the underlying cupy array and cause downstream accessors that are expecting a numpy array to fail?